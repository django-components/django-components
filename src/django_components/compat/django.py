from argparse import ArgumentParser
from typing import Any, Optional, Type

import django
from django.core.management.base import BaseCommand as DjangoCommand

from django_components.util.command import (
    CommandArg,
    ComponentCommand,
    _setup_command_arg,
    setup_parser_from_command,
)

# Django command arguments added to all commands
# NOTE: Many of these MUST be present for the command to work with Django
DJANGO_COMMAND_ARGS = [
    CommandArg(
        "--version",
        action="version",
        version=django.get_version(),
        help="Show program's version number and exit.",
    ),
    CommandArg(
        ["-v", "--verbosity"],
        default=1,
        type=int,
        choices=[0, 1, 2, 3],
        help=("Verbosity level; 0=minimal output, 1=normal output, 2=verbose output, " "3=very verbose output"),
    ),
    CommandArg(
        "--settings",
        help=(
            "The Python path to a settings module, e.g. "
            '"myproject.settings.main". If this isn\'t provided, the '
            "DJANGO_SETTINGS_MODULE environment variable will be used."
        ),
    ),
    CommandArg(
        "--pythonpath",
        help=("A directory to add to the Python path, e.g. " '"/home/djangoprojects/myproject".'),
    ),
    CommandArg(
        "--traceback",
        action="store_true",
        help="Raise on CommandError exceptions.",
    ),
    CommandArg(
        "--no-color",
        action="store_true",
        help="Don't colorize the command output.",
    ),
    CommandArg(
        "--force-color",
        action="store_true",
        help="Force colorization of the command output.",
    ),
    CommandArg(
        "--skip-checks",
        action="store_true",
        help="Skip system checks.",
    ),
]


def load_as_django_command(command: Type[ComponentCommand]) -> Type[DjangoCommand]:
    """
    Create a Django `Command` class from a `ComponentCommand` class.

    The created class can be used as a Django command by placing it in
    the `management/commands/` module.

    ```python
    # management/commands/mycommand.py
    from django_components.compat.django import load_as_django_command
    from myapp.commands.mycommand import MyCommand

    # NOTE: MUST be assigned to a variable named `Command`
    Command = load_as_django_command(MyCommand)
    ```
    """

    # Define a Command class that delegates to our command_class
    class Command(DjangoCommand):
        help = command.help

        def __init__(self) -> None:
            self._command = command()

        def create_parser(self, prog_name: str, subcommand: str, **kwargs: Any) -> ArgumentParser:
            parser = setup_parser_from_command(command)
            for arg in DJANGO_COMMAND_ARGS:
                _setup_command_arg(parser, arg.asdict())

            return parser

        # This is the entrypoint for the command - After argparser has resolved the args,
        # this is where we forward the args to the command handler.
        def handle(self, *args: Any, **options: Any) -> None:
            # Case: (Sub)command matched and it HAS handler
            resolved_command: Optional[ComponentCommand] = options.get("_command", None)
            if resolved_command and resolved_command.handle:
                resolved_command.handle(*args, **options)
                return

            # Case: (Sub)command matched and it DOES NOT have handler (e.g. subcommand used for routing)
            cmd_parser: Optional[ArgumentParser] = options.get("_parser", None)
            if cmd_parser:
                cmd_parser.print_help()
                return

            # Case: (Sub)command did not match - Print help for the main command
            self.print_help(self._command.name, "")

    Command.__doc__ = command.__doc__

    return Command
