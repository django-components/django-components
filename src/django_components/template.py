from contextlib import contextmanager
from typing import TYPE_CHECKING, Any, Dict, Generator, Optional, Type, Union, cast

from django.core.exceptions import ImproperlyConfigured
from django.template import Context, Origin, Template, TemplateDoesNotExist
from django.template.loader import get_template as django_get_template

from django_components.cache import get_template_cache
from django_components.template_loader import DjcLoader
from django_components.util.django_monkeypatch import is_template_cls_patched
from django_components.util.logger import trace_component_msg
from django_components.util.misc import get_import_path, get_module_info

if TYPE_CHECKING:
    from django_components.component import Component


# TODO_V1 - Remove, won't be needed once we remove `get_template_string()`, `get_template_name()`, `get_template()`
# Legacy logic for creating Templates from string
def cached_template(
    template_string: str,
    template_cls: Optional[Type[Template]] = None,
    origin: Optional[Origin] = None,
    name: Optional[str] = None,
    engine: Optional[Any] = None,
) -> Template:
    """
    DEPRECATED. Template caching will be removed in v1.

    Create a Template instance that will be cached as per the
    [`COMPONENTS.template_cache_size`](../settings#django_components.app_settings.ComponentsSettings.template_cache_size)
    setting.

    Args:
        template_string (str): Template as a string, same as the first argument to Django's\
            [`Template`](https://docs.djangoproject.com/en/5.2/topics/templates/#template). Required.
        template_cls (Type[Template], optional): Specify the Template class that should be instantiated.\
            Defaults to Django's [`Template`](https://docs.djangoproject.com/en/5.2/topics/templates/#template) class.
        origin (Type[Origin], optional): Sets \
            [`Template.Origin`](https://docs.djangoproject.com/en/5.2/howto/custom-template-backend/#origin-api-and-3rd-party-integration).
        name (Type[str], optional): Sets `Template.name`
        engine (Type[Any], optional): Sets `Template.engine`

    ```python
    from django_components import cached_template

    template = cached_template("Variable: {{ variable }}")

    # You can optionally specify Template class, and other Template inputs:
    class MyTemplate(Template):
        pass

    template = cached_template(
        "Variable: {{ variable }}",
        template_cls=MyTemplate,
        name=...
        origin=...
        engine=...
    )
    ```
    """  # noqa: E501
    template_cache = get_template_cache()

    template_cls = template_cls or Template
    template_cls_path = get_import_path(template_cls)
    engine_cls_path = get_import_path(engine.__class__) if engine else None
    cache_key = (template_cls_path, template_string, engine_cls_path)

    maybe_cached_template: Optional[Template] = template_cache.get(cache_key)
    if maybe_cached_template is None:
        template = template_cls(template_string, origin=origin, name=name, engine=engine)
        template_cache.set(cache_key, template)
    else:
        template = maybe_cached_template

    return template


@contextmanager
def prepare_component_template(
    component: "Component",
    template_data: Any,
) -> Generator[Optional[Template], Any, None]:
    context = component.context
    with context.update(template_data):
        template = _get_component_template(component)

        if template is None:
            # If template is None, then the component is "template-less",
            # and we skip template processing.
            yield template
            return

        if not is_template_cls_patched(template):
            raise RuntimeError(
                "Django-components received a Template instance which was not patched."
                "If you are using Django's Template class, check if you added django-components"
                "to INSTALLED_APPS. If you are using a custom template class, then you need to"
                "manually patch the class."
            )

        with _maybe_bind_template(context, template):
            yield template


def load_component_template(component_cls: Type["Component"], filepath: str) -> Template:
    if component_cls._template is not None:
        return component_cls._template

    # First try to load the filepath using our `DjcLoader`,
    # so we can pass in the component class to `DjcLoader` and so associate
    # the template with the component class via the `Origin` instance.
    #
    # If the template is NOT found, then we assume that the template
    # is to be loaded with other loaders, and hence we don't consider
    # this template to belong to any component.
    #
    # NOTE: This practically means that for some extension hooks like `on_template_preprocess()`
    #       to work, users MUST either:
    #       - Inline the template within the component class as `Component.template`
    #       - Set `Component.template_file` to filepath that is within our `COMPONENTS.dirs` / `COMPONENTS.app_dirs`
    #         and is thus is loaded by DjcLoader.
    #
    #       If users don't do this, then we don't know if the given template does or does not
    #       belong to a component, and thus we don't pre-process it.
    djc_loader = DjcLoader(None)
    try:
        template = djc_loader.get_template(filepath, component_cls=component_cls)
    except TemplateDoesNotExist:
        template = None

    # Otherwise, use Django's `get_template()` to load the template
    if template is None:
        template = _load_django_template(filepath)

    component_cls._template = template

    return template


# `_maybe_bind_template()` handles two problems:
#
# 1. Initially, the binding the template was needed for the context processor data
#    to work when using `RequestContext` (See `RequestContext.bind_template()` in e.g. Django v4.2 or v5.1).
#    But as of djc v0.140 (possibly earlier) we generate and apply the context processor data
#    ourselves in `Component._render_impl()`.
#
#    Now, we still want to "bind the template" by setting the `Context.template` attribute.
#    This is for compatibility with Django, because we don't know if there isn't some code that relies
#    on the `Context.template` attribute being set.
#
#    But we don't call `context.bind_template()` explicitly. If we did, then we would
#    be generating and applying the context processor data twice if the context was `RequestContext`.
#    Instead, we only run the same logic as `Context.bind_template()` but inlined.
#
#    The downstream effect of this is that if the user or some third-party library
#    uses custom subclass of `Context` with custom logic for `Context.bind_template()`,
#    then this custom logic will NOT be applied. In such case they should open an issue.
#
#    See https://github.com/django-components/django-components/issues/580
#    and https://github.com/django-components/django-components/issues/634
#
# 2. Not sure if I (Juro) remember right, but I think that with the binding of templates
#    there was also an issue that in *some* cases the template was already bound to the context
#    by the time we got to rendering the component. This is why we need to check if `context.template`
#    is already set.
#
#    The cause of this may have been compatibility with Django's `{% extends %}` tag, or
#    maybe when using the "isolated" context behavior. But not sure.
@contextmanager
def _maybe_bind_template(context: Context, template: Template) -> Generator[None, Any, None]:
    if context.template is not None:
        yield
        return

    # This code is taken from `Context.bind_template()` from Django v5.1
    context.template = template
    try:
        yield
    finally:
        context.template = None


def _get_component_template(component: "Component") -> Optional[Template]:
    trace_component_msg("COMP_LOAD", component_name=component.name, component_id=component.id, slot_name=None)

    # TODO_V1 - Remove, not needed once we remove `get_template_string()`, `get_template_name()`, `get_template()`
    template_sources: Dict[str, Optional[Union[str, Template]]] = {}

    # TODO_V1 - Remove `get_template_name()` in v1
    template_sources["get_template_name"] = component.get_template_name(component.context)

    # TODO_V1 - Remove `get_template_string()` in v1
    if hasattr(component, "get_template_string"):
        template_string_getter = getattr(component, "get_template_string")
        template_body_from_getter = template_string_getter(component.context)
    else:
        template_body_from_getter = None
    template_sources["get_template_string"] = template_body_from_getter

    # TODO_V1 - Remove `get_template()` in v1
    template_sources["get_template"] = component.get_template(component.context)

    # NOTE: `component.template` should be populated whether user has set `template` or `template_file`
    #       so we discern between the two cases by checking `component.template_file`
    if component.template_file is not None:
        template_sources["template_file"] = component.template_file
    else:
        template_sources["template"] = component.template

    # TODO_V1 - Remove this check in v1
    # Raise if there are multiple sources for the component template
    sources_with_values = [k for k, v in template_sources.items() if v is not None]
    if len(sources_with_values) > 1:
        raise ImproperlyConfigured(
            f"Component template was set multiple times in Component {component.name}."
            f"Sources: {sources_with_values}"
        )

    # Load the template based on the source
    if template_sources["get_template_name"]:
        template_name = template_sources["get_template_name"]
        template: Optional[Template] = _load_django_template(template_name)
        template_string: Optional[str] = None
    elif template_sources["get_template_string"]:
        template_string = template_sources["get_template_string"]
        template = None
    elif template_sources["get_template"]:
        # `Component.get_template()` returns either string or Template instance
        if hasattr(template_sources["get_template"], "render"):
            template = template_sources["get_template"]
            template_string = None
        else:
            template = None
            template_string = template_sources["get_template"]
    elif component.template or component.template_file:
        # If the template was loaded from `Component.template_file`, then the Template
        # instance was already created and cached in `Component._template`.
        #
        # NOTE: This is important to keep in mind, because the implication is that we should
        # treat Templates AND their nodelists as IMMUTABLE.
        if component.__class__._template is not None:
            template = component.__class__._template
            template_string = None
        # Otherwise user have set `Component.template` as string and we still need to
        # create the instance.
        else:
            template = _create_template_from_string(
                component,
                # NOTE: We can't reach this branch if `Component.template` is None
                cast(str, component.template),
                is_component_template=True,
            )
            template_string = None
    # No template
    else:
        template = None
        template_string = None

    # We already have a template instance, so we can return it
    if template is not None:
        return template
    # Create the template from the string
    elif template_string is not None:
        return _create_template_from_string(component, template_string)

    # Otherwise, Component has no template - this is valid, as it may be instead rendered
    # via `Component.on_render()`
    return None


def _create_template_from_string(
    component: "Component",
    template_string: str,
    is_component_template: bool = False,
) -> Template:
    # Generate a valid Origin instance.
    # When an Origin instance is created by Django when using Django's loaders, it looks like this:
    # ```
    # {
    #   'name': '/path/to/project/django-components/sampleproject/calendarapp/templates/calendarapp/calendar.html',
    #   'template_name': 'calendarapp/calendar.html',
    #   'loader': <django.template.loaders.app_directories.Loader object at 0x10b441d90>
    # }
    # ```
    #
    # Since our template is inlined, we will format as `filepath::ComponentName`
    #
    # ```
    # /path/to/project/django-components/src/calendarapp/calendar.html::Calendar
    # ```
    #
    # See https://docs.djangoproject.com/en/5.2/howto/custom-template-backend/#template-origin-api
    _, _, module_filepath = get_module_info(component.__class__)
    origin = Origin(
        name=f"{module_filepath}::{component.__class__.__name__}",
        template_name=None,
        loader=None,
    )

    _set_origin_component(origin, component.__class__)

    if is_component_template:
        template = Template(template_string, name=origin.template_name, origin=origin)
        component.__class__._template = template
    else:
        # TODO_V1 - `cached_template()` won't be needed as there will be only 1 template per component
        #           so we will be able to instead use `template_cache` to store the template
        template = cached_template(
            template_string=template_string,
            name=origin.template_name,
            origin=origin,
        )

    return template


# When loading a template, use Django's `get_template()` to ensure it triggers Django template loaders
# See https://github.com/django-components/django-components/issues/901
#
# This may raise `TemplateDoesNotExist` if the template doesn't exist.
# See https://docs.djangoproject.com/en/5.2/ref/templates/api/#template-loaders
# And https://docs.djangoproject.com/en/5.2/ref/templates/api/#custom-template-loaders
#
# TODO_v3 - Instead of loading templates with Django's `get_template()`,
#       we should simply read the files directly (same as we do for JS and CSS).
#       This has the implications that:
#       - We would no longer support Django's template loaders
#       - Instead if users are using template loaders, they should re-create them as djc extensions
#       - We would no longer need to set `TEMPLATES.OPTIONS.loaders` to include
#         `django_components.template_loader.Loader`
def _load_django_template(template_name: str) -> Template:
    return django_get_template(template_name).template


def _set_origin_component(origin: Origin, component_cls: Type["Component"]) -> None:
    origin.component_cls = component_cls


def _get_origin_component(origin: Origin) -> Optional[Type["Component"]]:
    return getattr(origin, "component_cls", None)
