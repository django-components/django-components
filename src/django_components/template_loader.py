"""
Template loader that loads templates from each Django app's "components" directory.
"""

from pathlib import Path
from typing import TYPE_CHECKING, Generator, List, Optional, Type

from django.core.exceptions import SuspiciousFileOperation
from django.template import Origin, Template, TemplateDoesNotExist
from django.template.loaders.filesystem import Loader as FilesystemLoader
from django.utils._os import safe_join

from django_components.util.loader import get_component_dirs

if TYPE_CHECKING:
    from django_components.component import Component


class DjcLoader(FilesystemLoader):
    def get_dirs(self, include_apps: bool = True) -> List[Path]:
        """
        Prepare directories that may contain component files:

        Searches for dirs set in `COMPONENTS.dirs` settings. If none set, defaults to searching
        for a "components" app. The dirs in `COMPONENTS.dirs` must be absolute paths.

        In addition to that, also all apps are checked for `[app]/components` dirs.

        Paths are accepted only if they resolve to a directory.
        E.g. `/path/to/django_project/my_app/components/`.

        `BASE_DIR` setting is required.
        """
        return get_component_dirs(include_apps)

    # Same as `FilesystemLoader.get_template()` from Django v5.1, except optionally
    # accepts a Component class that the template belongs to.
    # This is used by `load_component_template()` to associate the template with the component class.
    def get_template(
        self,
        template_name: str,
        skip: Optional[List[Origin]] = None,
        component_cls: Optional[Type["Component"]] = None,
    ) -> Template:
        tried = []

        for origin in self.get_template_sources(template_name, component_cls=component_cls):
            if skip is not None and origin in skip:
                tried.append((origin, "Skipped to avoid recursion"))
                continue

            try:
                contents = self.get_contents(origin)
            except TemplateDoesNotExist:
                tried.append((origin, "Source does not exist"))
                continue
            else:
                return Template(
                    contents,
                    origin,
                    origin.template_name,
                    self.engine,
                )

        raise TemplateDoesNotExist(template_name, tried=tried)

    # Same as `FilesystemLoader.get_template_sources()` from Django v5.1, except optionally
    # accepts a Component class that the template belongs to.
    # This is used by `load_component_template()` to associate the template with the component class.
    def get_template_sources(
        self,
        template_name: str,
        component_cls: Optional[Type["Component"]] = None,
    ) -> Generator[Origin, None, None]:
        for template_dir in self.get_dirs():
            try:
                name = safe_join(template_dir, template_name)
            except SuspiciousFileOperation:
                # The joined path was located outside of this template_dir
                # (it might be inside another one, so this isn't fatal).
                continue

            origin = Origin(
                name=name,
                template_name=template_name,
                loader=self,
            )

            if component_cls:
                from django_components.template import _set_origin_component

                _set_origin_component(origin, component_cls)

            yield origin

    # Same as `FilesystemLoader.get_contents()` from Django v5.1, except that it defaults
    # to `file_charset` UTF-8 if engine is not set.
    # This is used by `load_component_template()` so we don't need an instance of Django's `Engine`
    # to call `DjcLoader.get_template()`.
    def get_contents(self, origin: Origin) -> str:
        encoding = self.engine.file_charset if self.engine else "utf-8"
        try:
            with open(origin.name, encoding=encoding) as fp:
                return fp.read()
        except FileNotFoundError:
            raise TemplateDoesNotExist(origin)


# NOTE: Django's template loaders have the pattern of using the `Loader` class name.
#       However, this then makes it harder to track and distinguish between different loaders.
#       So internally we use the name `DjcLoader` instead.
#       But for public API we use the name `Loader` to match Django.
Loader = DjcLoader
