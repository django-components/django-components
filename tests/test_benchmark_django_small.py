# NOTE: This file is used for benchmarking. Before editing this file,
# please read through these:
# - `benchmarks/README`
# - https://github.com/django-components/django-components/pull/999

from pathlib import Path
from typing import Dict, Literal, NamedTuple, Optional, Union

import django
from django.conf import settings
from django.template import Context, Template

from django_components import types

# DO NOT REMOVE - See https://github.com/django-components/django-components/pull/999
# ----------- IMPORTS END ------------ #

# This variable is overridden by the benchmark runner
CONTEXT_MODE: Literal["django", "isolated"] = "isolated"

if not settings.configured:
    settings.configure(
        BASE_DIR=Path(__file__).resolve().parent,
        INSTALLED_APPS=["django_components"],
        TEMPLATES=[
            {
                "BACKEND": "django.template.backends.django.DjangoTemplates",
                "DIRS": [
                    "tests/templates/",
                    "tests/components/",  # Required for template relative imports in tests
                ],
                "OPTIONS": {
                    "builtins": [
                        "django_components.templatetags.component_tags",
                    ]
                },
            }
        ],
        COMPONENTS={
            "template_cache_size": 128,
            "autodiscover": False,
            "context_behavior": CONTEXT_MODE,
        },
        MIDDLEWARE=["django_components.middleware.ComponentDependencyMiddleware"],
        DATABASES={
            "default": {
                "ENGINE": "django.db.backends.sqlite3",
                "NAME": ":memory:",
            }
        },
        SECRET_KEY="secret",
        ROOT_URLCONF="django_components.urls",
    )
    django.setup()
else:
    settings.COMPONENTS["context_behavior"] = CONTEXT_MODE

#####################################
#
# IMPLEMENTATION START
#
#####################################

templates_cache: Dict[int, Template] = {}


def lazy_load_template(template: str) -> Template:
    template_hash = hash(template)
    if template_hash in templates_cache:
        return templates_cache[template_hash]
    else:
        template_instance = Template(template)
        templates_cache[template_hash] = template_instance
        return template_instance


#####################################
# RENDER ENTRYPOINT
#####################################


def gen_render_data():
    data = ButtonData(
        href="https://example.com",
        disabled=False,
        variant="primary",
        type="button",
        attrs={
            "class": "py-2 px-4",
        },
        slot_content="Click me!",
    )
    return data


def render(data: "ButtonData"):
    # Render
    result = button(Context(), data)
    return result


#####################################
# THEME
#####################################

ThemeColor = Literal["default", "error", "success", "alert", "info"]
ThemeVariant = Literal["primary", "secondary"]

VARIANTS = ["primary", "secondary"]


class ThemeStylingUnit(NamedTuple):
    """
    Smallest unit of info, this class defines a specific styling of a specific
    component in a specific state.

    E.g. styling of a disabled "Error" button.
    """

    color: str
    """CSS class(es) specifying color"""
    css: str = ""
    """Other CSS classes not specific to color"""


class ThemeStylingVariant(NamedTuple):
    """
    Collection of styling combinations that are meaningful as a group.

    E.g. all "error" variants - primary, disabled, secondary, ...
    """

    primary: ThemeStylingUnit
    primary_disabled: ThemeStylingUnit
    secondary: ThemeStylingUnit
    secondary_disabled: ThemeStylingUnit


class Theme(NamedTuple):
    """Class for defining a styling and color theme for the app."""

    default: ThemeStylingVariant
    error: ThemeStylingVariant
    alert: ThemeStylingVariant
    success: ThemeStylingVariant
    info: ThemeStylingVariant


_secondary_btn_styling = "ring-1 ring-inset"

theme = Theme(
    default=ThemeStylingVariant(
        primary=ThemeStylingUnit(
            color="bg-blue-600 text-white hover:bg-blue-500 focus-visible:outline-blue-600 transition"
        ),
        primary_disabled=ThemeStylingUnit(color="bg-blue-300 text-blue-50 focus-visible:outline-blue-600 transition"),
        secondary=ThemeStylingUnit(
            color="bg-white text-gray-800 ring-gray-300 hover:bg-gray-100 focus-visible:outline-gray-600 transition",
            css=_secondary_btn_styling,
        ),
        secondary_disabled=ThemeStylingUnit(
            color="bg-white text-gray-300 ring-gray-300 focus-visible:outline-gray-600 transition",
            css=_secondary_btn_styling,
        ),
    ),
    error=ThemeStylingVariant(
        primary=ThemeStylingUnit(color="bg-red-600 text-white hover:bg-red-500 focus-visible:outline-red-600"),
        primary_disabled=ThemeStylingUnit(color="bg-red-300 text-white focus-visible:outline-red-600"),
        secondary=ThemeStylingUnit(
            color="bg-white text-red-600 ring-red-300 hover:bg-red-100 focus-visible:outline-red-600",
            css=_secondary_btn_styling,
        ),
        secondary_disabled=ThemeStylingUnit(
            color="bg-white text-red-200 ring-red-100 focus-visible:outline-red-600",
            css=_secondary_btn_styling,
        ),
    ),
    alert=ThemeStylingVariant(
        primary=ThemeStylingUnit(color="bg-amber-500 text-white hover:bg-amber-400 focus-visible:outline-amber-500"),
        primary_disabled=ThemeStylingUnit(color="bg-amber-100 text-orange-300 focus-visible:outline-amber-500"),
        secondary=ThemeStylingUnit(
            color="bg-white text-amber-500 ring-amber-300 hover:bg-amber-100 focus-visible:outline-amber-500",
            css=_secondary_btn_styling,
        ),
        secondary_disabled=ThemeStylingUnit(
            color="bg-white text-orange-200 ring-amber-100 focus-visible:outline-amber-500",
            css=_secondary_btn_styling,
        ),
    ),
    success=ThemeStylingVariant(
        primary=ThemeStylingUnit(color="bg-green-600 text-white hover:bg-green-500 focus-visible:outline-green-600"),
        primary_disabled=ThemeStylingUnit(color="bg-green-300 text-white focus-visible:outline-green-600"),
        secondary=ThemeStylingUnit(
            color="bg-white text-green-600 ring-green-300 hover:bg-green-100 focus-visible:outline-green-600",
            css=_secondary_btn_styling,
        ),
        secondary_disabled=ThemeStylingUnit(
            color="bg-white text-green-200 ring-green-100 focus-visible:outline-green-600",
            css=_secondary_btn_styling,
        ),
    ),
    info=ThemeStylingVariant(
        primary=ThemeStylingUnit(color="bg-sky-600 text-white hover:bg-sky-500 focus-visible:outline-sky-600"),
        primary_disabled=ThemeStylingUnit(color="bg-sky-300 text-white focus-visible:outline-sky-600"),
        secondary=ThemeStylingUnit(
            color="bg-white text-sky-600 ring-sky-300 hover:bg-sky-100 focus-visible:outline-sky-600",
            css=_secondary_btn_styling,
        ),
        secondary_disabled=ThemeStylingUnit(
            color="bg-white text-sky-200 ring-sky-100 focus-visible:outline-sky-600",
            css=_secondary_btn_styling,
        ),
    ),
)


def get_styling_css(
    variant: Optional["ThemeVariant"] = None,
    color: Optional["ThemeColor"] = None,
    disabled: Optional[bool] = None,
):
    """
    Dynamically access CSS styling classes for a specific variant and state.

    E.g. following two calls get styling classes for:
    1. Secondary error state
    1. Secondary alert disabled state
    2. Primary default disabled state
    ```py
    get_styling_css('secondary', 'error')
    get_styling_css('secondary', 'alert', disabled=True)
    get_styling_css(disabled=True)
    ```
    """
    variant = variant or "primary"
    color = color or "default"
    disabled = disabled if disabled is not None else False

    color_variants: ThemeStylingVariant = getattr(theme, color)

    if variant not in VARIANTS:
        raise ValueError(f'Unknown theme variant "{variant}", must be one of {VARIANTS}')

    variant_name = variant if not disabled else f"{variant}_disabled"
    styling: ThemeStylingUnit = getattr(color_variants, variant_name)

    css = f"{styling.color} {styling.css}".strip()
    return css


#####################################
# BUTTON
#####################################

button_template_str: types.django_html = """
    {# Based on buttons from https://tailwindui.com/components/application-ui/overlays/modals #}

    {% if is_link %}
    <a
        href="{{ href }}"
        {% html_attrs attrs class=btn_class class="no-underline" %}
    >
    {% else %}
    <button
        type="{{ type }}"
        {% if disabled %} disabled {% endif %}
        {% html_attrs attrs class=btn_class %}
    >
    {% endif %}

        {{ slot_content }}

    {% if is_link %}
    </a>
    {% else %}
    </button>
    {% endif %}
"""


class ButtonData(NamedTuple):
    href: Optional[str] = None
    link: Optional[bool] = None
    disabled: Optional[bool] = False
    variant: Union["ThemeVariant", Literal["plain"]] = "primary"
    color: Union["ThemeColor", str] = "default"
    type: Optional[str] = "button"
    attrs: Optional[dict] = None
    slot_content: Optional[str] = ""


def button(context: Context, data: ButtonData):
    common_css = (
        "inline-flex w-full text-sm font-semibold"
        " sm:mt-0 sm:w-auto focus-visible:outline-2 focus-visible:outline-offset-2"
    )
    if data.variant == "plain":
        all_css_class = common_css
    else:
        button_classes = get_styling_css(data.variant, data.color, data.disabled)  # type: ignore[arg-type]
        all_css_class = f"{button_classes} {common_css} px-3 py-2 justify-center rounded-md shadow-sm"

    is_link = not data.disabled and (data.href or data.link)

    all_attrs = {**(data.attrs or {})}
    if data.disabled:
        all_attrs["aria-disabled"] = "true"

    with context.push(
        {
            "href": data.href,
            "disabled": data.disabled,
            "type": data.type,
            "btn_class": all_css_class,
            "attrs": all_attrs,
            "is_link": is_link,
            "slot_content": data.slot_content,
        }
    ):
        return lazy_load_template(button_template_str).render(context)


#####################################
#
# IMPLEMENTATION END
#
#####################################


# DO NOT REMOVE - See https://github.com/django-components/django-components/pull/999
# ----------- TESTS START ------------ #
# The code above is used also used when benchmarking.
# The section below is NOT included.

from django_components.testing import djc_test  # noqa: E402

@djc_test
def test_render(snapshot):
    data = gen_render_data()
    rendered = render(data)
    assert rendered == snapshot
