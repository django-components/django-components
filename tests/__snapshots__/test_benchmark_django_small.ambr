# serializer version: 1
# name: test_render
  '''
  
      
  
      
      <a
          href="https://example.com"
          class="py-2 px-4 bg-blue-600 text-white hover:bg-blue-500 focus-visible:outline-blue-600 transition inline-flex w-full text-sm font-semibold sm:mt-0 sm:w-auto focus-visible:outline-2 focus-visible:outline-offset-2 px-3 justify-center rounded-md shadow-sm no-underline"
      >
      
  
          Click me!
  
      
      </a>
      
  
  '''
# ---
